Second price auction based on ContractHost

Chris-Hibbert opened the pull request:
A checkpoint for visibility.

Chris-Hibbert committed:
Second Price auction

code is drafty; tests are started.ˆ

Chris-Hibbert committed:
checkpoint

Chris-Hibbert committed:
checkpoint

Chris-Hibbert committed:
checkpoint

Chris-Hibbert committed:
checkpoint

Chris-Hibbert committed:
The auction works now. Clean-up remains.

Chris-Hibbert committed:
The auction works and a simple test passes.

I'm currently working on verification of the contract and seats.

Chris-Hibbert committed:
The autction works now. There is cleap yet to do.

Validate terms in vat-bidders.

Chris-Hibbert committed:
verifying terms.

Chris-Hibbert committed:
progress on testing, but inscrutable failure in vouch()

Chris-Hibbert committed:
debugging

Chris-Hibbert committed:
Tests now pass. Time to clean up debugging stuff.  Thanks, Kate!

Chris-Hibbert committed:
removed debugger statements, and excess logging.

Chris-Hibbert committed:
removed calls to tick() in auction code

Chris-Hibbert committed:
final cleanups

Chris-Hibbert committed:
more cleanups

katelynsills commented on test/unitTests/core/test-issuers.js:
```diff
@@ -39,6 +39,26 @@ test('split good amounts', t => {
   }
 });
 
+test('split good w/different amounts', t => {
```
Would it be possible to take this test out of this PR since we have a SwingSet based test for the bug on master now?

Chris-Hibbert replied:
sure.

Chris-Hibbert committed:
drop duplicated test

katelynsills reviewed (changes requested):
I'm thinking that we probably don't want to add `agencyEscrow.js` into `core/ `since it will be confusing to have multiple escrows, and this one seems specific to `auction.js`. I think I'd be ok adding both to `more/` mostly as is.

Chris-Hibbert committed:
rename left/right to currencyAmount/goodsAmount

Chris-Hibbert committed:
drop test.only

Chris-Hibbert committed:
use has() instead of get() for boolean result

Chris-Hibbert committed:
convert bestPrice/secondPrice to originalOffer/finalPrice for generality in agencyEscrow

Chris-Hibbert commented on core/agencyEscrow.js:
```diff
@@ -0,0 +1,119 @@
+        const { issuer: currencyIssuer } = currencyAmount.label;
+        const overbidAmount = E(currencyIssuer).makeAmount(
+          bestPrice - secondPrice,
+        );
```
I renamed bestPrice/secondPrice to originalOffer/finalPrice to make it clear that it might be useful in other cases where the bidders might not spend everything they offered. 

Is this enough to make it clear that it could be used for other things?

katelynsills replied:
Hmm, I think it probably couldn't easily be reused, even with the renaming change. If it lives under `/more/auction/` with `auction.js`, I'm ok with using it as an example and learning device. I think it shouldn't go under core/ if there is already an `escrow.js` there.

Chris-Hibbert replied:
done.

Chris-Hibbert commented on core/agencyEscrow.js:
```diff
@@ -0,0 +1,119 @@
+// received for the transaction.
+const agencyEscrow = {
+  start: (terms, inviteMaker) => {
+    const { left: currencyAmount, right: goodsAmount } = terms;
```
Done.

Chris-Hibbert commented on test/swingsetTests/auction/bootstrap.js:
```diff
@@ -0,0 +1,138 @@
+        }
+        log(`@@ schedule task for:${deadline}, currently: ${ticks} @@`);
+        const result = makePromise();
+        if (!schedule.get(deadline)) {
```
Done

Chris-Hibbert commented on test/swingsetTests/auction/bootstrap.js:
```diff
@@ -0,0 +1,138 @@
+      tick(msg) {
+        ticks += 1;
+        log(`@@ tick:${ticks}${msg ? `: ${msg}` : ''} @@`);
+        if (schedule.get(ticks)) {
```
Done.

Chris-Hibbert committed:
extract manualTimer to tools/

Chris-Hibbert commented on test/swingsetTests/auction/bootstrap.js:
```diff
@@ -0,0 +1,138 @@
+import { makeUniAssayConfigMaker } from '../../../core/config/uniAssayConfig';
+
+function build(E, log) {
+  function buildManualTimer(startValue = 0) {
```
done.

Chris-Hibbert committed:
move auction.js and agencyEscrow.js out of core.

katelynsills commented on more/auction/agencyEscrow.js:
```diff
@@ -0,0 +1,119 @@
+        const wonGoodsPayment = E(goodsIssuer).claimAll(goodsPayment, 'wins');
+        const { issuer: currencyIssuer } = currencyAmount.label;
+        const overbidAmount = E(currencyIssuer).makeAmount(
+          originalOffer - finalPrice,
```
I think we shouldn't hardcode a Nat Strategy here. We could use `assay.without` or get the strategy from the issuer.

katelynsills reviewed (dismissed):
I think there's a couple of changes but once those are in, looks good to me!

Chris-Hibbert committed:
checkpoint

Chris-Hibbert committed:
changes from review

Chris-Hibbert committed:
added all requested changes from reviews

Chris-Hibbert committed:
eslint issue

Chris-Hibbert committed:
checkpoint

Chris-Hibbert committed:
ERTP renamings, step 1. Make the tests pass

Chris-Hibbert committed:
More renaming cleanups.

Removed checkInstallation from agencyEscrow, since it doesn't currently hand out any invites.

Chris-Hibbert commented:
I've now integrated with the ERTP renamings.

Chris-Hibbert committed:
make  uniDescOpsConfig work and use makeNatExtentOps

Chris-Hibbert commented:
Please take another look at this.

katelynsills commented on core/contractHost.js:
```diff
@@ -15,6 +15,33 @@ import { makeInviteConfig } from './config/inviteConfig';
 import { makeMint } from './mint';
 import makePromise from '../util/makePromise';
 
+function makeCollect(E, log) {
```
Is this just moved up?

Chris-Hibbert replied:
One minor change: 
```
 const doneP = allSettled(results);	
    Promise.resolve(doneP).then(([wins, refs]) => {
```
became
```
    allSettled(results).then(([wins, refs]) => {
```

katelynsills replied:
If it was moved, would it be possible to move it back where it was so this minor change stands out? I'm seeing a wall of addition/subtractions. Is it just git?

Chris-Hibbert replied:
Yep. done. 

I've seen tools that could do a better job of displaying changes in moved code, but it's not common.

katelynsills replied:
looks good!

katelynsills reviewed (dismissed):
Looks good to me if the contractHost.js changes are reverted or clarified

katelynsills commented on test/swingsetTests/auction/vat-bidders.js:
```diff
@@ -0,0 +1,81 @@
+
+          const allegedInviteBalanceP = E(bidderSeatPaymentP).getBalance();
+          E.resolve(
+            allComparable(harden([termsP, allegedInviteBalanceP])),
```
sounds good!

katelynsills commented on more/auction/auction.js:
```diff
@@ -0,0 +1,167 @@
+      return E(currencyIssuer)
+        .claimAll(paymentP)
+        .then(currencyPaymentP => {
+          return E(currencyPaymentP)
```
Either way is fine. I generally prefer leaving it out when I can but that's just my preference

Chris-Hibbert commented:
I can't tell if my responses to your earlier comments went out. Are there remaining issues with contractHost?

Chris-Hibbert committed:
put collect back at the bottom of contractHost.js

katelynsills reviewed (approved):
lgtm!

erights commented on more/auction/auction.js:
```diff
@@ -0,0 +1,198 @@
+          return;
+        }
+        cancelExcept(bestBidderP);
+        const bestBidAgencySeatP = agencySeatsP.get(bestBidderP);
```
Be sure you're looking up presences, which have reliable identity, rather than promises, which do not.

erights commented on more/auction/auction.js:
```diff
@@ -0,0 +1,198 @@
+        cancelExcept(bestBidderP);
+        const bestBidAgencySeatP = agencySeatsP.get(bestBidderP);
+        const paidAssetDescP = E(bestBidAgencySeatP).consummateDeal(
+          bestPrice,
```
I see. But you already know that it offered the best price. The consumeDeal API is better without this parameter.

erights commented on more/auction/auction.js:
```diff
@@ -0,0 +1,198 @@
+      // allows the holder to make offers and ensures that the money will be
+      // returned unless they get the goods.
+      makeBidderSeat() {
+        const seats = E(agencyEscrowInstallationP).spawn(escrowTerms);
```
Need to check that the auction is still open.

erights commented on more/auction/auction.js:
```diff
@@ -0,0 +1,198 @@
+          getRefund() {
+            return E(buyerSeatP).getRefund();
+          },
+          // Can the bidder cancel? Not currently.
```
Good. It is a second price auction after all ;).

erights commented on more/auction/auction.js:
```diff
@@ -0,0 +1,198 @@
+          },
+          // Can the bidder cancel? Not currently.
+        });
+        const bidderId = `bidder${bidderSeatCount}`;
```
I suppose "Id" is short for "identifier" so I shouldn't complain. But I only use `Id` for unforgeable things. Perhaps I should consistently use `token` or something instead?

erights commented:
@Chris-Hibbert Can we close this as obsolete? I doubt we will want to invest effort to create new contracts on contractHost (though we should maintain the ones already there as tests).

Chris-Hibbert commented:
Obsolete.
